sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/BaseAuthorizationTokenProvider.java

```diff
@@ -237,38 +239,44 @@ private String generateKeyAuthorizationSignatureNew(RequestVerb verb, String res
 
         String authResourceId = getAuthorizationResourceIdOrFullName(resourceType, resourceIdValue);
         String payLoad = generateMessagePayload(verb, authResourceId, resourceType, headers);
-        Mac mac = this.getMacInstance();
-        byte[] digest = mac.doFinal(payLoad.getBytes(StandardCharsets.UTF_8));
-        String authorizationToken = Utils.encodeBase64String(digest);
-        String authtoken = AUTH_PREFIX + authorizationToken;
-        return HttpUtils.urlEncode(authtoken);
+
+        MacPool.ReUsableMac macInstance = this.getReUseableMacInstance();
+
+        try {
+            byte[] digest = macInstance.get().doFinal(payLoad.getBytes(StandardCharsets.UTF_8));
+            String authorizationToken = Utils.encodeBase64String(digest);
+            String authtoken = AUTH_PREFIX + authorizationToken;
+            return HttpUtils.urlEncode(authtoken);
+        }
+        finally {
+            macInstance.close();
+        }
     }
 
-    private Mac getMacInstance() {
-        int masterKeyLatestHashCode = this.credential.getKey().hashCode();
-
-        //  Master key has changed, or this is the first time we are getting mac instance
-        if (masterKeyLatestHashCode != this.masterKeyHashCode) {
-            byte[] masterKeyBytes = this.credential.getKey().getBytes(StandardCharsets.UTF_8);
-            byte[] masterKeyDecodedBytes = Utils.Base64Decoder.decode(masterKeyBytes);
-            SecretKey signingKey = new SecretKeySpec(masterKeyDecodedBytes, "HMACSHA256");
-            try {
-                Mac macInstance = Mac.getInstance("HMACSHA256");
-                macInstance.init(signingKey);
-                //  Update the master key hash code
-                this.masterKeyHashCode = masterKeyLatestHashCode;
-                return macInstance;
-            } catch (NoSuchAlgorithmException | InvalidKeyException e) {
-                throw new IllegalStateException(e);
-            }
-        } else {
-            //  Master key hasn't changed, return the cloned mac instance
-            try {
-                return (Mac)this.macInstance.clone();
-            } catch (CloneNotSupportedException e) {
-                throw new IllegalStateException(e);
+    private MacPool.ReUsableMac getReUseableMacInstance() {
+
+        // Java == operator is reference equals not content
+        // leveraging reference comparison avoid hash computation
+        if (this.currentCredentialKey != this.credential.getKey()) {
```

Comment:
SpotBugs cribbing, @kushagraThapar, @moderakh how to suppress this warning?

Reply:
@kirankumarkolli  - you need to update this file with the proper spotbug issue which is failing here : https://github.com/Azure/azure-sdk-for-java/blob/master/eng/code-quality-reports/src/main/resources/spotbugs/spotbugs-exclude.xml